planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+
+
+## Unit Testing Approach
+
+If we introduce the runtime code generation back into Marten, that's unfortunately a non-trivial "cold start" testing issue. To soften that, I suggest we get a lot more aggressive with reusable [xUnit.Net class fixtures](https://xunit.net/docs/shared-context) between tests to reuse generated code between tests, cut way down on the sheer number of database calls by not having to frequently check the schema configuration, and other `DocumentStore` overhead. 
+
+A couple other points about this:
+
+* We need to create more unique document types so we're not having to use different configurations for the same document type. This would enable more reuse inside the testing runtime
+* Be aggressive with separate schemas for different configurations
+* We could possibly turn on xUnit.net parallel test running to speed up the test cycles
+
+
+## Document Metadata
+
+* Use the configuration and tests from [pull request for Document Metadata](https://github.com/JasperFx/marten/pull/1364), but use the Lamar-backed dynamic code generation from the previous section to pull this off.
+* I think we could expand the document metadata to allow for user defined properties like "user id" or "transaction id" much the same way we'll do for the EventStore metadata. We'd need to think about how we extend the document tables and how metadata is attached to a document session
+
+
+## Project / Assembly Restructuring
+
+* Pull everything to do with Schema object generation and difference detection to a separate library (`IFeatureSchema`, `ISchemaObject`, etc.). Mostly to clean out the main library, but also because this code could easily be reused outside of Marten. Separating it out might make it easier to test and extend that functionality, which is something that occasionally gets requested. There's also the possibility of further breaking that into abstractions and implementations for the long run of getting us ready for Sql Server or other database engine support
+
+* *Possibly* pull the ADO.Net helper code like `CommandBuilder` and the extension methods into a small helper library somewhere else (I'm nominating the [Baseline](https://jasperfx.github.io/baseline) repository). This code is copied around to other projects as it is, and it's another way of getting stuff out of the main library and the test suite.
+
+* More in the event store section, but let's get the event store functionality into its own library (or possibly a couple other libraries for async projection support).
+
+## F# Improvements
+
+WE NEED TO HAVE AN F# SUBCOMMITTEE TO HELP HERE. 
+
+## HostBuilder Integration
+
+If he's okay with this, I'd vote to bring Joona-Pekka Kokko's ASP.Net Core integration library into the main repository and make that the officially blessed and documented recipe for integrating Marten into .Net Core applications based on the `HostBuilder` in .Net Core. I suppose we could also multi-target `IWebHostBuilder` for ASP.Net Core 2.*.
```

Comment:
Yes, we'll need to also add more documentation and samples for that.

Reply:
I'm fine with this.